here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/AbstractTask.java

```diff
@@ -398,63 +402,61 @@ protected final void unlock() {
-    final AtomicLong limit =
-        new AtomicLong((long) Runtime.getRuntime().availableProcessors() * thresholdPerProcessor);
-    final long LIMIT = limit.get();
-    final double authorThreshold = (double) percentageThresholdPerPrinciple / 100;
+    final long LIMIT = AbstractTask.limit.get();
     lockAndRequireNew();
     try {
-      do {
-        final long threadCount = AbstractTask.threadCount.get();
-        assert threadCount >= 0L;
-        String author = context.getAuthor();
-        if (author != null) {
-          AtomicInteger authorUsage = AbstractTask.authorUsageMap.getOrDefault(author, new AtomicInteger(0));
-          int authorLimit = (int) (LIMIT * authorThreshold);
-          if (!internal && authorUsage.get() >= authorLimit) {
-            log.info(
-                "NAKSHA_ERR_REQ_LIMIT_4_PRINCIPAL - [Request Limit breached for Principal => appId,author,limit,crtValue] - ReqLimitForPrincipal {} {} {} {}",
-                context.getAppId(),
-                author,
-                authorLimit,
-                authorUsage);
-            throw new TooManyTasks(0, (long) (authorLimit), false);
-          }
-        }
-        if (!internal && threadCount >= LIMIT) {
+      long myIncrementedValue = AbstractTask.threadCount.updateAndGet(l -> {
+        if (!internal && AbstractTask.threadCount.get() >= LIMIT) {
           log.info(
               "NAKSHA_ERR_REQ_LIMIT_4_INSTANCE - [Request Limit breached for Instance => appId,author,limit,crtValue] - ReqLimitForInstance {} {} {} {}",
               context.getAppId(),
-              author,
+              getAuthorName(),
               LIMIT,
               threadCount);
-          throw new TooManyTasks(LIMIT, 0, true);
+          String errorMessage = "Maximum number of concurrent tasks reached for instance (" + LIMIT + ")";
+          throw new TooManyTasks(errorMessage);
         }
-        if (AbstractTask.threadCount.compareAndSet(threadCount, threadCount + 1)) {
-          if (author != null) {
-            AbstractTask.incrementAuthorUsage(author);
-          }
-          try {
-            state.set(State.START);
-            final Future<RESULT> future = threadPool.submit(this::init_and_execute);
-            return future;
-          } catch (RejectedExecutionException e) {
-            throw new TooManyTasks();
-          } catch (Throwable t) {
-            AbstractTask.threadCount.decrementAndGet();
-            if (author != null) {
-              AbstractTask.decrementAuthorUsage(author);
-            }
-            log.atError()
-                .setMessage("Unexpected exception while trying to fork a new thread")
-                .setCause(t)
-                .log();
-            throw new RuntimeException("Internal error while forking new worker thread", t);
-          }
-        }
-        // Conflict, two threads concurrently try to fork.
-      } while (true);
+        return l + 1;
+      });
+      assert myIncrementedValue >= 0L;
```

Comment:
We need to change order here: Do not reserve a slot (worker thread) before you have evaluated that the principal even has the right to do this. Otherwise, an evil principal will still affect us badly, because he will always reserve a slot (work) and later, ones we detect that the principal does not even have the right to acquire a slot, because he already exceeds the limits, have to release the slot again, when this happens fast enough, nobody ever will be able to do anything, except for the evil principal, because he will reserve all slots (workers), even while they will be released soon.

Reply:
Fixed.